src/pdftoraw.c

```diff
@@ -0,0 +1,189 @@
+#include <gio/gio.h>
+#include <poppler.h>
+#include <gdk/gdk.h>
+
+#define DPI 150.0
+
+static struct option long_options[] = {
+    {"file",     required_argument, 0,  'f' },
+    {"pages",    no_argument,       0,  'p' },
+    {"test",     no_argument,       0,  't' },
+    {"width",    required_argument, 0,  'w' },
+    {"height",   required_argument, 0,  'W' },
+    {"raw",      required_argument, 0,  'r' },
+    {"help",     no_argument,       0,  'h' },
+    {0,          0,                 0,  0   }
+};
+
+static void
+print_usage(void) {
+    printf("Usage:\tpdftoraw  --filename=<filename>  --raw=<num-page>\n");
+    printf("\tpdftoraw  --filename=<filename>  --pages\n");
+    printf("\tpdftoraw  --filename=<filename>  --width=<num-page>\n");
+    printf("\tpdftoraw  --filename=<filename>  --height=<num-page>\n");
+    printf("\tpdftoraw  --help\n");
+}
+
+static int
+pdf_number_pages_get(PopplerDocument *doc)
+{
+    return poppler_document_get_n_pages(doc);
+}
+
+static unsigned char*
+pdf_page_raw_get(PopplerDocument *doc,
+                 int page_nr)
+{
+    cairo_surface_t *s = NULL;
+    cairo_t *cr = NULL;
+    unsigned char *data = NULL;
+    GdkPixbuf *pix = NULL;
+    double width = 0;
+    double height = 0;
+    int w = 0;
+    int h = 0;
+
+    PopplerPage *page = poppler_document_get_page(doc, page_nr);
+    poppler_page_get_size(page, &width, &height);
+    width = DPI * width / 72.0;
+    height = DPI * height / 72.0;
+    w = (int)ceil(width);
+    h = (int)ceil(height);
+    s = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, w, h);
+    cr = cairo_create(s);
+    cairo_scale (cr, DPI/72.0, DPI/72.0);
+    cairo_save (cr);
+    poppler_page_render_for_printing(page, cr);
+    cairo_restore (cr);
+    pix = gdk_pixbuf_get_from_surface(s, 0, 0, w, h);
+    g_object_unref(page);
+    data = gdk_pixbuf_get_pixels (pix);
+    fwrite(data, 1, (w * h * 4), stdout);
+    return data;
+}
+
+static int
+pdf_page_width_get(PopplerDocument *doc,
+                   int page_nr)
+{
+    double width = 0;
+    int w = 0;
+
+    PopplerPage *page = poppler_document_get_page(doc, page_nr);
+    poppler_page_get_size(page, &width, NULL);
+    width = DPI * width / 72.0;
+    w = (int)ceil(width);
+    g_object_unref(page);
+    return w;
+}
+
+static int
+pdf_page_height_get(PopplerDocument *doc,
+                    int page_nr)
+{
+    double height = 0;
+    int h = 0;
+
+    PopplerPage *page = poppler_document_get_page(doc, page_nr);
+    poppler_page_get_size(page, NULL, &height);
+    height = DPI * height / 72.0;
+    h = (int)ceil(height);
+    g_object_unref(page);
+    return h;
+}
+
+int
+main(int argc, char **argv)
+{
+    PopplerDocument *doc = NULL;
+    GError *err = NULL;
+    GFile *in = NULL;
+    char *filename = NULL;
+    int long_index =0;
+    int page = 0;
+    int opt= 0;
+    int test = 0, raw = 0, npages = 0, width = 0, height = 0;
+
+    while ((opt = getopt_long(argc, argv,"f:tpw:W:r:h",
```

Comment:
GLib has its own functions for parsing args. See `g_option_context_parse()`.

Reply:
Fixed by 007bd452e1ebb4b